### PR TITLE
feat: kotlin multipart

### PR DESCRIPTION
**Pull request by byawitz**
## What does this PR do?

(Provide a description of what this PR does.)

## Test Plan

(Write your test plan here. If you changed any code, please provide us with clear instructions on how you verified your changes work.)

## Related PRs and Issues

(If this PR is related to any other PR or resolves any issue or related to any issue link all related PR and issues here.)

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

(Write your answer here.)

### CODE REVIEW

File: mock-server/app/http.php [hunk 1]
```diff
@@ -317,12 +317,11 @@
     ->param('x', '', new Text(100), 'Sample string param')
     ->param('y', '', new Integer(true), 'Sample numeric param')
     ->param('z', null, new ArrayList(new Text(256), APP_LIMIT_ARRAY_PARAMS_SIZE), 'Sample array param')
-    ->param('payload', [], new File(), 'Sample file param', skipValidation: true)
+    ->param('file', [], new File(), 'Sample file param', skipValidation: true)
```
Comment (Meldiron) [inline mock-server/app/http.php:320]: Is this nessessary? Can we make tests pass without this, like other SDKs?

File: templates/kotlin/src/main/kotlin/io/appwrite/Client.kt.twig [hunk 2]
```diff
@@ -557,11 +558,17 @@ class Client @JvmOverloads constructor(
                     it.resume(true as T)
                     return
                 }
+                if (response.headers["content-type"]?.contains("multipart/form-data") == true) {
```
Comment (Meldiron) [inline templates/kotlin/src/main/kotlin/io/appwrite/Client.kt.twig:561]:
```suggestion
                if (response.headers["content-type"]?.contains("multipart/form-data")) {
```

File: templates/kotlin/src/main/kotlin/io/appwrite/extensions/TypeExtensions.kt.twig [hunk 3]
```diff
@@ -1,8 +1,72 @@
-package {{ sdk.namespace | caseDot }}.extensions
+package io.appwrite.extensions
```
Comment (Meldiron) [inline templates/kotlin/src/main/kotlin/io/appwrite/extensions/TypeExtensions.kt.twig:1]: Is this change nessessary?

File: templates/kotlin/src/main/kotlin/io/appwrite/extensions/TypeExtensions.kt.twig [hunk 4]
```diff
@@ -1,8 +1,72 @@
...
 
+import io.appwrite.models.Payload
```
Comment (Meldiron) [inline templates/kotlin/src/main/kotlin/io/appwrite/extensions/TypeExtensions.kt.twig:3]: Can we use namespace placeholder?

File: templates/kotlin/src/main/kotlin/io/appwrite/extensions/TypeExtensions.kt.twig [hunk 5]
```diff
@@ -1,8 +1,72 @@
...
+        "\$createdAt" to "",
+        "\$updatedAt" to "",
+        "\$permissions" to emptyList<String>(),
+        "functionId" to "",
+        "trigger" to "",
+        "status" to "",
+        "requestMethod" to "",
+        "requestPath" to "",
+        "requestHeaders" to emptyList<HashMap<String, Any>>(),
+        "statusCode" to 0,
```
Comment (Meldiron) [inline templates/kotlin/src/main/kotlin/io/appwrite/extensions/TypeExtensions.kt.twig:27]: This should be responseStatusCode

File: templates/kotlin/src/main/kotlin/io/appwrite/models/Payload.kt.twig [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+import java.nio.file.Files
+import java.nio.file.Paths
+
+class Payload private constructor() {
+
+    lateinit var path: String
+    lateinit var filename: String
+    lateinit var mimeType: String
+    lateinit var sourceType: String
+    lateinit var data: Any
```
Comment (Meldiron) [inline templates/kotlin/src/main/kotlin/io/appwrite/models/Payload.kt.twig:15]: I see some complexity here. Is all of this needed?

For example path - do we even need to remember this? Do we send it to API?
Or sourceType - can't `data` just always be ByteArray, and in fromX methods we always convert to ByteArray? API doesn't care what type it is
mimeType - I think API doesn't need it. it only needs file name

File: templates/kotlin/src/main/kotlin/io/appwrite/models/Payload.kt.twig [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+                ?: URLConnection.guessContentTypeFromName(filename)
+                        ?: ""
+            sourceType = "file"
+        }
+
+        fun fromPath(path: String): Payload = fromFile(File(path)).apply {
+            sourceType = "path"
+        }
+
+        fun fromBinary(bytes: ByteArray, filename: String = "", mimeType: String = "") = Payload().apply {
```
Comment (Meldiron) [inline templates/kotlin/src/main/kotlin/io/appwrite/models/Payload.kt.twig:56]: mimeType doesnt follow structure we decided. Let me know if it's necessary for implementation, i which case we might need to apply it to all SDKs. otherwise, lets follow this, and update all methods accordingly:

```
function toBinary(): bytes[]
function toJson(): JSON
function toString(): string
function toFile(path: string): void

static function fromBinary(bytes: bytes[], name?: string): Payload
static function fromFile(file: File, name?: string): Payload // Name truly optional in Storage
static function fromJson(object: JSON, name?: string): Payload
static function fromString(text: string, name?: string): Payload
```

File: tests/languages/kotlin/Tests.kt [hunk 8]
```diff
@@ -133,6 +135,14 @@ class ServiceTest {
             )
             writeToFile(url)
 
+            // Multipart tests
+            val mp = general.multipart()
+
+            writeToFile((mp as Map<String, Any>)["x"] as String)
```
Comment (Meldiron) [inline tests/languages/kotlin/Tests.kt:141]: What is the type of `mp`? Will developers also need to do all of those complex casts?

File: src/SDK/Language/Kotlin.php [hunk 9]
```diff
@@ -191,15 +192,18 @@ public function getParamDefault(array $param): string
...
         $output = '';
 
         if (empty($example) && $example !== 0 && $example !== false) {
             switch ($type) {
+                case self::TYPE_PAYLOAD:
+                    $output .= 'payload.fromString("<BODY>")';
+                    break;
                 case self::TYPE_FILE:
-                    $output .= 'InputFile.fromPath("file.png")';
+                    $output .= 'payload.fromPath("file.png")';
```
Comment (Meldiron) [inline src/SDK/Language/Kotlin.php:206]:
```suggestion
                    $output .= 'Payload.fromPath("file.png")';
```

File: src/SDK/Language/Kotlin.php [hunk 10]
```diff
@@ -240,8 +244,11 @@ public function getParamExample(array $param): string
                 case self::TYPE_BOOLEAN:
                     $output .= ($example) ? 'true' : 'false';
                     break;
+                case self::TYPE_PAYLOAD:
+                    $output .= 'payload.fromString("<BODY>")';
```
Comment (Meldiron) [inline src/SDK/Language/Kotlin.php:248]:
```suggestion
                    $output .= 'Payload.fromString("<BODY>")';
```

File: src/SDK/Language/Kotlin.php [hunk 11]
```diff
@@ -480,6 +487,10 @@ protected function getModelType(array $definition, array $spec, string $generic
 
     protected function getPropertyType(array $property, array $spec, string $generic = 'T'): string
     {
+        if (str_contains($property['description'] ?? '', 'HTTP response body. This will return empty unless execution')) {
```
Comment (Meldiron) [inline src/SDK/Language/Kotlin.php:490]: Can we find a better check than this? Maybe based on action + param name

File: templates/kotlin/src/main/kotlin/io/appwrite/models/Payload.kt.twig [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+import io.appwrite.extensions.gson
+import java.io.File
+import java.net.URLConnection
+import java.nio.file.Files
+import java.nio.file.Paths
+
+class Payload private constructor() {
+
+    lateinit var path: String
+    lateinit var filename: String
```
Comment (Meldiron) [inline templates/kotlin/src/main/kotlin/io/appwrite/models/Payload.kt.twig:12]: I feel like I am missing something. How does filename get to API?

Comment (byawitz) [reply templates/kotlin/src/main/kotlin/io/appwrite/models/Payload.kt.twig:12]: [here](https://github.com/appwrite/sdk-generator/blob/4d2f3eaa9fcad9cc73478276ace2cd98392bbf13/templates/kotlin/src/main/kotlin/io/appwrite/Client.kt.twig#L349) for example

File: templates/kotlin/src/main/kotlin/io/appwrite/models/Payload.kt.twig (hunk 6 above)
Comment (byawitz) [reply templates/kotlin/src/main/kotlin/io/appwrite/models/Payload.kt.twig:15]: The path is used to access the [file](https://github.com/appwrite/sdk-generator/blob/4d2f3eaa9fcad9cc73478276ace2cd98392bbf13/templates/kotlin/src/main/kotlin/io/appwrite/Client.kt.twig#L332)
[Source type](https://github.com/appwrite/sdk-generator/blob/4d2f3eaa9fcad9cc73478276ace2cd98392bbf13/templates/kotlin/src/main/kotlin/io/appwrite/Client.kt.twig#L342) is used to fetch the file before push and the code is in the client, same as for mimeType

File: tests/languages/kotlin/Tests.kt (hunk 8 above)
Comment (byawitz) [reply tests/languages/kotlin/Tests.kt:141]: It's not needed no, only

File: templates/kotlin/src/main/kotlin/io/appwrite/Client.kt.twig [hunk 13]
```diff
@@ -549,6 +550,15 @@ class Client @JvmOverloads constructor(
                         return
                     }
                 }
+                println(response.headers["content-type"])
+                if (response.headers["content-type"]?.contains("multipart/form-data") == true) {
+                    val binaryBody = response.body!!.bytes()
+                    val body = String(binaryBody)
```
Comment (christyjacob4) [inline templates/kotlin/src/main/kotlin/io/appwrite/Client.kt.twig:556]: is this the reason we get inconsistency in the hashes

**Review by Meldiron — APPROVED** [review]

File: src/SDK/Language/Kotlin.php [hunk 14]
```diff
@@ -480,6 +490,14 @@ protected function getModelType(array $definition, array $spec, string $generic
 
     protected function getPropertyType(array $property, array $spec, string $generic = 'T'): string
     {
+        if ($property['name'] == 'responseBody') {
```
Comment (vermakhushboo) [inline src/SDK/Language/Kotlin.php:493]: Do we need this special check here?

File: templates/kotlin/src/main/kotlin/io/appwrite/extensions/TypeExtensions.kt.twig [hunk 15]
```diff
@@ -1,8 +1,107 @@
...
 import kotlin.reflect.typeOf
 
 inline fun <reified T : Any> classOf(): Class<T> {
     return (typeOf<T>().classifier!! as KClass<T>).java
-}
\ No newline at end of file
+}
+
+fun String.fromMultiPart(binaryBody: ByteArray): Map<String, Any> {
+    val match = Regex("(-+\\w+)--").find(this) ?: return emptyMap()
```
Comment (vermakhushboo) [inline templates/kotlin/src/main/kotlin/io/appwrite/extensions/TypeExtensions.kt.twig:12]: Is regex slow? Can we use a faster check here?

File: templates/kotlin/src/main/kotlin/io/appwrite/models/Payload.kt.twig [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+    lateinit var sourceType: String
+    lateinit var data: Any
+    var mimeType: String? = null
```
Comment (vermakhushboo) [inline templates/kotlin/src/main/kotlin/io/appwrite/models/Payload.kt.twig:15]: Do we need `mimetype` and `sourceType`?

File: templates/kotlin/src/main/kotlin/io/appwrite/extensions/TypeExtensions.kt.twig (hunk 15 above)
Comment (Meldiron) [reply templates/kotlin/src/main/kotlin/io/appwrite/extensions/TypeExtensions.kt.twig:12]: In ideal world we would use well tested library for this, but to release quickly, we use this snippet.

Stability for us should be fine, as tests pass.

Speed for us should be fine, we only have a few keys. Could be a bit slower for 20MB payload, but regex looks pretty simple so I wouldn't be too worried

File: templates/kotlin/src/main/kotlin/io/appwrite/models/Payload.kt.twig [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+            this.filename = filename
+            data = bytes
+            sourceType = "bytes"
+        }
+
+        fun fromString(string: String) = fromBinary(string.toByteArray())
+
+        fun fromJson(data: Any) = fromString(gson.toJson(data))
+
+        fun fromFileObject(file: File, name: String = "") = Payload().apply {
```
Comment (vermakhushboo) [inline templates/kotlin/src/main/kotlin/io/appwrite/models/Payload.kt.twig:62]: Can we call this `fromFile` for consistency with other languages?

File: templates/kotlin/src/main/kotlin/io/appwrite/models/Payload.kt.twig (hunk 16 above)
Comment (Meldiron) [reply templates/kotlin/src/main/kotlin/io/appwrite/models/Payload.kt.twig:15]: Its used in Client.kt, so I would say yes. Looking at Payload methods, mimeType is always set internally. So if we ever need to improve it, we can as it wont be user-facing.

sourceType is used to do memory-optimal chunk upload.

File: tests/KotlinJava17Test.php [hunk 18]
```diff
@@ -17,7 +17,7 @@ class KotlinJava17Test extends Base
         'chmod +x tests/sdks/kotlin/gradlew',
     ];
     protected string $command =
-        'docker run --network="mockapi" -v $(pwd):/app -w /app/tests/sdks/kotlin openjdk:17-jdk-slim sh -c "./gradlew test -q && cat result.txt"';
+        'docker run --network="mockapi" --rm -v $(pwd):/app -w /app/tests/sdks/kotlin openjdk:17-jdk-slim sh -c "./gradlew test -q && cat result.txt"';
```
Comment (vermakhushboo) [inline tests/KotlinJava17Test.php:20]: Can you explain why was this change required?

Comment (Meldiron) [reply tests/KotlinJava17Test.php:20]: It's just cleanup, so I can safely re-run tests without having deleting container manually. I see the same in other tests too.

**Review by vermakhushboo — APPROVED** [review]

**Review by loks0n — APPROVED** [review]
